extensions/flags/src/Api/Controller/ListFlagsController.php

```diff
@@ -28,26 +29,57 @@ class ListFlagsController extends AbstractListController
         'post.discussion'
     ];
 
+    public function __construct(
+        protected UrlGenerator $url
+    ) {
+    }
+
     protected function data(ServerRequestInterface $request, Document $document): iterable
     {
         $actor = RequestUtil::getActor($request);
-        $include = $this->extractInclude($request);
 
         $actor->assertRegistered();
 
         $actor->read_flags_at = Carbon::now();
         $actor->save();
 
-        $flags = Flag::whereVisibleTo($actor)
-            ->latest('flags.created_at')
-            ->groupBy('post_id')
-            ->get();
+        $limit = $this->extractLimit($request);
+        $offset = $this->extractOffset($request);
+        $include = $this->extractInclude($request);
 
         if (in_array('post.user', $include)) {
             $include[] = 'post.user.groups';
         }
 
-        $this->loadRelations($flags, $include);
+        $primaries = Flag::whereVisibleTo($actor)
+            ->groupBy('post_id')
+            ->orderBy('created_at', 'DESC')
+            ->skip($offset)
+            ->take($limit + 1);
+
+        $flags = Flag::whereVisibleTo($actor)
+            ->select('flags.*')
+            ->joinSub($primaries, 'p', 'flags.id', '=', 'p.id')
+            ->latest()
+            ->get();
```

Comment:
I don't understand the two queries, what does it do and why not a simple paginated query?

Reply:
I just copied how it was done from `NotificationRepository`. When I tried a single query it wouldn't load more past the first 20.

Reply:
Oh I see. Yea this querying logic is unique to notifications, so what you want to do is recover the old simple flag query and in addition apply a ->limit($limit + 1)->offset($offset) and it should be good to go

Reply:
Alrighty!